pl_bolts/datamodules/experience_source.py

```diff
@@ -29,174 +31,265 @@ def __iter__(self) -> Iterable:
-        self.gamma = gamma
+
+        self.pool = env if isinstance(env, (list, tuple)) else [env]
+        self.exp_history_queue = deque()
+
         self.n_steps = n_steps
-        self.n_step_buffer = deque(maxlen=n_steps)
+        self.total_steps = []
+        self.states = []
+        self.histories = []
+        self.cur_rewards = []
+        self.cur_steps = []
+        self.iter_idx = 0
 
-    def step(self, device: torch.device) -> Tuple[Experience, float, bool]:
-        """
-        Takes an n-step in the environment
+        self._total_rewards = []
+
+        self.init_envs()
+
+    def runner(self, device: torch.device) -> Tuple[Experience]:
+        """Experience Source iterator yielding Tuple of experiences for n_steps. These come from the pool
+        of environments provided by the user.
+
+        Args:
+            device: current device to be used for executing experience steps
 
         Returns:
-            Experience
+            Tuple of Experiences
         """
-        exp = self.n_step(device)
+        while True:
+            # get actions for all envs
+            actions = self.env_actions(device)
 
-        while len(self.n_step_buffer) < self.n_steps:
-            self.n_step(device)
+            # step through each env
+            for env_idx, (env, action) in enumerate(zip(self.pool, actions)):
 
-        reward, next_state, done = self.get_transition_info()
-        first_experience = self.n_step_buffer[0]
-        multi_step_experience = Experience(
-            first_experience.state, first_experience.action, reward, done, next_state
-        )
+                exp = self.env_step(env_idx, env, action)
+                history = self.histories[env_idx]
+                history.append(exp)
+                self.states[env_idx] = exp.new_state
 
-        return multi_step_experience, exp.reward, exp.done
+                self.update_history_queue(env_idx, exp, history)
 
-    def n_step(self, device: torch.device) -> Experience:
+                # Yield all accumulated history tuples to model
+                while self.exp_history_queue:
+                    yield self.exp_history_queue.popleft()
+
+            self.iter_idx += 1
+
+    def update_history_queue(self, env_idx, exp, history) -> None:
         """
-        Takes a  single step in the environment and appends it to the n-step buffer
+        Updates the experience history queue with the lastest experiences. In the event of an experience step is in
+        the done state, the history will be incrementally appended to the queue, removing the tail of the history
+        each time.
+        Args:
+            env_idx: index of the environment
+            exp: the current experience
+            history: history of experience steps for this environment
+        """
+        # If there is a full history of step, append history to queue
+        if len(history) == self.n_steps:
+            self.exp_history_queue.append(tuple(history))
```

Comment:
does has to e tuple?

Reply:
yeah, the history is sliced later on and you can't slice a deque